[4.1] Misc improvements for Task\TaskOption

ditsuke opened the pull request:
### Summary of Changes
- Add `@property-reads` docs for better IDE/static analysis support.
- Refactor `TaskOption::type` to `TaskOption::id`

### Testing Instructions
- Create a new Scheduled Task with a short interval-based schedule (1 minute).
- Let it run organically at planned time (should be a minute after created).
- Run it manually from the button in the Scheduled Tasks manager.


### Actual result BEFORE applying this Pull Request
- Everything works as expected; but:  


- TaskOption::type doesn't make sense.
- Poor IDE support for $taskOption->{property} since they're accessed only through TaskOption::__get().

### Expected result AFTER applying this Pull Request
- Everything works as expected; but:  

- TaskOption::id makes more sense.
- IDE doesn't complain about valid $taskOption->{property{ reads.


### Documentation Changes Required
\-

ditsuke committed:
Add @property-read doc for Task\TaskOption

Better doc and IDE support.

ditsuke committed:
Refactor TaskOption::type to TaskOption::id

Because ID makes sense, "type" does not.

ditsuke committed:
Fix code-style

bembelimen committed:
Merge branch '4.1-dev' into improve-taskoption-misc

PhilETaylor commented on administrator/components/com_scheduler/src/Task/TaskOption.php:
```diff
@@ -20,6 +20,11 @@
  * argument pairs as argument. Internally, the TaskOption object generates the routine title and description from the
  * language constant prefix.
  *
+ * @property-read  string $desc             The routine description.
+ * @property-read  string $id               The routine ID.
+ * @property-read  string $langConstPrefix  The routine's language constant prefix.
+ * @property-read  string $title            The routine title.
+ *
  * @since  4.1.0
  */
```
property-read should come after since, and a blank line.

ditsuke committed:
Fix docblock order

ditsuke committed:
Merge remote-tracking branch 'ditsuke/improve-taskoption-misc' into improve-taskoption-misc

ditsuke committed:
Merge branch '4.1-dev' into improve-taskoption-misc

ditsuke committed:
Merge branch '4.1-dev' into improve-taskoption-misc

ditsuke committed:
Merge branch '4.1-dev' into improve-taskoption-misc

ditsuke commented:
Tests, please!

ditsuke committed:
Merge branch '4.1-dev' into improve-taskoption-misc

Commit:
Merge branch '4.1-dev' into improve-taskoption-misc

Quy commented:
Create a new task.

> Cannot use object of type Joomla\CMS\Object\CMSObject as array 

> JROOT\administrator\components\com_scheduler\src\Traits\TaskPluginTrait.php:214

ditsuke committed:
Fix TaskOption->id accesses in TaskPluginTrait

These couldn't be covered by static-analysis backed refactor
since we don't have any type information in this method.

ditsuke commented:
> Create a new task.
> 
> > Cannot use object of type Joomla\CMS\Object\CMSObject as array
> 
> > JROOT\administrator\components\com_scheduler\src\Traits\TaskPluginTrait.php:214

@Quy should be resolved with [0525674](https://github.com/joomla/joomla-cms/pull/36516/commits/05256744d1ced2a622e79b62745b7f7d70ba156c). Requesting a retest.

Quy commented:
Manually running the task.

![36516](https://user-images.githubusercontent.com/368084/152257219-073c76c3-630a-48cd-bdc1-08a895f5ea48.png)

ditsuke commented:
> Manually running the task.
> 
> ![36516](https://user-images.githubusercontent.com/368084/152257219-073c76c3-630a-48cd-bdc1-08a895f5ea48.png)

@Quy I'm unable to reproduce. Can you please share the configuration of the failing task?

ditsuke committed:
Fix execution exception (complete refactoring)

Outdated JS build! :(

ditsuke commented:
@Quy apologies, my efforts to test were being hampered by an outdated JS build and changes from #36518! We should be good to go now with 56d5b7b 😅

Quy commented:
I have tested this item :white_check_mark: successfully on 56d5b7b50c5e888fc6e79fc3b4326a4686634e55<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36516">issues.joomla.org/tracker/joomla-cms/36516</a>.</sub>

jwaisner commented:
I have tested this item :white_check_mark: successfully on 56d5b7b50c5e888fc6e79fc3b4326a4686634e55<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36516">issues.joomla.org/tracker/joomla-cms/36516</a>.</sub>

jwaisner commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36516">issues.joomla.org/tracker/joomla-cms/36516</a>.</sub>

bembelimen commented:
I guess this will break all 3rd party extension, using ```->type``` already, so you need to add the fallback into the magic get method (if "type" use "id") and deprecate it the same moment to make it work.

ditsuke committed:
Add back TaskOption::type with deprecated status

This will help with b/c compat and allow 3PDs to update their
extensions to TaskOption::id.

ditsuke commented:
> I guess this will break all 3rd party extension, using ```->type``` already, so you need to add the fallback into the magic get method (if "type" use "id") and deprecate it the same moment to make it work.

@bembelimen added back with deprecated status.  Please review.

ditsuke committed:
Merge branch '4.1-dev' into improve-taskoption-misc

richard67 commented:
@bembelimen Does it need 2 tests again? Or is it ok to keep RTC and you review?

bembelimen commented:
We should use the log class instead throwing a PHP depreciation:

https://github.com/joomla/joomla-cms/blob/4.1-dev/libraries/src/Application/CMSApplication.php#L418

richard67 commented:
> We should use the log class instead throwing a PHP depreciation:
> 
> https://github.com/joomla/joomla-cms/blob/4.1-dev/libraries/src/Application/CMSApplication.php#L418
> 
> Then review is enough

@ditsuke Could you implement that suggestion? Thanks in advance.

ditsuke committed:
Convert deprecation error to log

ditsuke committed:
Merge branch '4.1-dev' into improve-taskoption-misc

ditsuke commented:
@richard67 thanks for tagging. Implemented with ef13678.

bembelimen commented:
Thx

ditsuke commented:
thanks
